docs/resource-tuning-recommendations.md

```diff
@@ -0,0 +1,122 @@
+---
+title: Resource Tuning Recommendations
+resourceType: Documentation
+personas: 
+  - Developer
+tags:
+  - openshift
+  - resources
+  - optimization
+  - devops
+  - jenkins
+description: How to tune your application resources.
+---
+
+# Application Resource Tuning
+
+As touched upon in the [Resiliency Guidelines](https://developer.gov.bc.ca/Resiliency-Guidelines#a-correctly-resourced-app), deploying applications with appropriate CPU and Memory [*Requests and Limits*](https://docs.openshift.com/container-platform/3.11/admin_guide/overcommit.html#requests-and-limits) is critical to ensure: 
+- Resource availability for your own applications
+- Resource availability for other tenant applications
+
+While [*Resource Quota's*](https://docs.openshift.com/container-platform/3.11/dev_guide/compute_resources.html#dev-quotas) are quite generous, it's important that this quota is seen as a tool to allow tenants enough resources to temporarily burst usage for experimentation, rather than an upper limit of consistent use. The platform is not sized to support every tenant fully utilizing their *Resource Quota*. 
+
+**Resource Requests**  
+Resource requests are guaranteed and reserved for the pod. *Scheduling decisions are made based on the request* to ensure that a node has enough capacity available to meet the requested value.
+
+**Resource Limits**  
+Resource limits set an upper limit of what a pod can burst to if the resources are available on the node. 
+
+## Setting Requests and Limits
+:exclamation: **If you set a resource Limit, you should also set a resource Request.** Otherwise the Request will match the Limit. For example, a Deployment with *no* defined cpu request and a definted cpu limit of 1 core will *result in a pod with a request of 1 cpu and a limit of 1 cpu*.
+
+**General Guidelines**  
+:ballot_box_with_check: Set requests and limits.  
+:ballot_box_with_check: Set requests to the *minimum* of what your application needs.  
+:ballot_box_with_check: Set limits to a resonable burstable number of what a single pod should support.   
+:ballot_box_with_check: Use horizontal pod autoscalers where possible, rather than large cpu and memory limits.   
+
+
+### Resource FAQ's
+
+**Q: What happens if I don't specify a limit or a request?**  
```

Comment:
Do we want to discuss BestEffort pods were the Req/Limits are set to zero? Or just clarify that unset is different than zero?

Reply:
I've added another question and answer in here to address this. It's also in the resiliency guide, though I think the resiliency guide should ultimately reference this document as well (since resourcing in itself it's own topic)

Reply:
Actually, this doesn't work. Users cannot set their pods to 0. The defaults will kick in.

Reply:
Works for me? Or is that just because I'm a cluster-admin?

```console
$ oc describe dc django-psql-example | grep -A5 Limits
    Limits:
      cpu:      0
      memory:   0
    Requests:
      cpu:      0
      memory:   0
$ oc describe quota compute-best-effort-quota
Name:       compute-best-effort-quota
Namespace:  steve-besteffort
Scopes:     BestEffort
 * Matches all pods that do not have resource requirements set. These pods have a best effort quality of service.
Resource  Used  Hard
--------  ----  ----
pods      1     3
$ oc describe node <node> | grep -E 'CPU Requests|django-psql-example-1-46jk8'
  Namespace                           Name                                            CPU Requests  CPU Limits  Memory Requests  Memory Limits
  steve-besteffort                    django-psql-example-1-46jk8                     0 (0%)        0 (0%)      0 (0%)           0 (0%)
```

Reply:
Well, thanks for digging deeper @sbarre-esit. A couple things have come to light
1. I fixed the patch command in this example (as it was slightly off)
2. I had tested through the Web Console; when setting `0` for the resource request/limit, it simply removes the value altogether from the deployment and the defaults continue to apply to the pods when the spin up. If, however, you use the CLI to patch it to `0`, if you use the Web Console to "edit and save" the resource configuration and leave the values at `0`, it will again remove them and apply defaults. 
3. When patched from the CLI, the `0` will stay in effect and actually apply the BestEffort QoS policy. Setting this value on the CLI with `oc edit` also functions as expected. 

```
$ oc describe dc jenkins1 | grep -A5 Limits
$ oc describe quota compute-best-effort-quota
Name:       compute-best-effort-quota
Namespace:  c81e6h-tools
Scopes:     BestEffort
 * Matches all pods that do not have resource requirements set. These pods have a best effort quality of service.
Resource  Used  Hard
--------  ----  ----
pods      0     3
$ oc patch dc/jenkins1 -p '{"spec": {"template": {"spec": {"containers":[{"name":"jenkins", "resources":{"requests": {"cpu":"0", "memory":"0"}, "limits": {"cpu":"0", "memory":"0"}}}]}}}}'
deploymentconfig.apps.openshift.io/jenkins1 patched
[sheastewart@sheawei ~]$ oc describe dc jenkins1 | grep -A5 Limits
    Limits:
      cpu:	0
      memory:	0
    Requests:
      cpu:	0
      memory:	0
$ oc describe quota compute-best-effort-quota
Name:       compute-best-effort-quota
Namespace:  c81e6h-tools
Scopes:     BestEffort
 * Matches all pods that do not have resource requirements set. These pods have a best effort quality of service.
Resource  Used  Hard
--------  ----  ----
pods      1     3 
```

Reply:
I feel like the latest updates have resolved this conversation @sbarre-esit . Thoughts?